### ISSUE TITLE
Admin filter showing tasks that can still have active downloads

### ISSUE THREAD

**Issue by rwrzesien** [golemfactory/concent#329]
Extracted from https://github.com/golemfactory/concent/issues/152#issuecomment-384696044. This task depends on #660 and #661.

Add a new filter in the admin panel. It should show subtasks that are either in an active state or are in `RESULT UPLOADED` state while current time <= `TaskToCompute.deadline` + SVT. We'll leave the decision whether to wait for downloads or not up to the person shutting down the service.

To implement this you either need to resolve #141 first or add a `deadline` column to the `Subtask` table. If you do the latter, make sure that it's updated every time `task_to_compute` is updated (like RCT in #174).

**Comment by cameel** [comment]
We can do this even without #141. Just add `deadline` column on the `Subtask` table (and don't forget to update the #139 blueprint).

**Comment by cameel** [comment]
Also, let's split this into two issues. The first point is much more important than the second one.

**Comment by cameel** [comment]
I have created a separate task for the second point (#335) and updated the description.

**Comment by dybi** [comment]
@cameel, as discussed, another filed in `Subtask` model will be needed:
`result_package_size` from `ReportComputedTask`

**Comment by cameel** [comment]
@dybi About a query that finds the tasks that still have active downloads.

I was thinking about something along these lines:
``` python
from django.db.models import F
from django.db.models import Func
from django.db.models import Manager
from django.db.models import Model
from django.db.models import Q


class SubtaskWithTimingColumnsManager(Manager):
    def get_queryset(self):
        bytes_per_sec   = settings.MINIMUM_UPLOAD_RATE << 10
        download_time   = Func(F('result_package_size') / float(bytes_per_sec), function='CEIL'),

        mdt             = F('maximum_download_time')
        ttc_dt          = F('task_to_compute__timestamp')
        subtask_dt      = F('computation_deadline')
        subtask_timeout = Func('second', subtask_dt - ttc_dt, function='DATE_PART')

        return super().get_queryset().annotate(
            maximum_download_time     = settings.DOWNLOAD_LEADIN_TIME + download_time,
            subtask_verification_time = (4 * settings.CONCENT_MESSAGING_TIME) + (3 * mdt) + (0.5 * subtask_timeout),
        )


class Subtask(Model):
    # Current field definitions
    ...
    ...
    objects_with_timing_columns = SubtaskWithTimingColumnsManager()
    ...


subtasks_with_active_downloads = Subtask.objects_with_timing_columns.filter(
    Q(current_time <= F('computation_deadline') + F('maximum_download_time'))
)
```

I have not tried to run it yet and I can see a few places which could be wrong (more details below), but these are problems or the "am I using it right" kind rather than "is it even possible to do it in a query" kind. You can see that these are just two simple expressions composed of constants, column names, basic arithmetic operators and simple SQL functions that you can use on columns in a SELECT query. Each of them could be a one-liner but I wanted to keep the intermediate variables like in the original functions so that they're easier to compare.

A few remarks:
- This will obviously only work after you add `computation_deadline` and `result_package_size` columns to the `Subtask` table. The first should match `task_to_compute.deadline`, the second `report_task_to_compute.size`.
- `download_time` may require casts to/from int/float. I assumed that PostgreSQL will convert it to the right type automatically but if not, just use a [Cast](https://docs.djangoproject.com/en/1.11/ref/models/database-functions/#cast) expression.
- The result of `DATE_PART()` might require a cast too.
- Here's the reference for PostgreSQL's [mathematical functions](https://www.postgresql.org/docs/current/static/functions-math.html) and [date/time functions](https://www.postgresql.org/docs/current/static/functions-datetime.html) in case you need to tweak it.
- The code above creates a custom model manager so that you can use `objects_with_timing_columns` rather than the default `objects` and get model instances annotated with the extra columns right away. You're most likely familiar with it but if not, see [Django docs > Managers > Modifying a manager’s initial QuerySet](https://docs.djangoproject.com/en/2.0/topics/db/managers/#modifying-a-manager-s-initial-queryset). You could also simply put it in a class method on a model and skip defining the manager but I think the solution with a model manager is nicer to use.
- To get access to these annotated columns in the admin panel, override [ModelAdmin.get_queryset()](https://docs.djangoproject.com/en/2.0/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_queryset) and use this custom model manager.

**Comment by cameel** [comment]
I have created additional tasks as prerequisites for this one: #660 and #661.
